src/OpenTelemetry/Metrics/MeterProviderSdk.cs

```diff
@@ -94,16 +96,26 @@ internal sealed class MeterProviderSdk : MeterProvider
                 {
                     if (meterSourcesToSubscribe.ContainsKey(instrument.Meter.Name))
                     {
-                        var index = Interlocked.Increment(ref this.metricIndex);
-                        if (index >= MaxMetrics)
+                        lock (this.instrumentCreationLock)
                         {
-                            // Log that all measurements are dropped from this instrument.
-                        }
-                        else
-                        {
-                            var metric = new Metric(instrument, temporality);
-                            this.metrics[index] = metric;
-                            listener.EnableMeasurementEvents(instrument, metric);
+                            if (this.metricStreamNames.ContainsKey(instrument.Name))
+                            {
+                                // log and ignore this instrument.
+                                return;
+                            }
+
+                            var index = Interlocked.Increment(ref this.metricIndex);
+                            if (index >= MaxMetrics)
+                            {
+                                // Log that all measurements are dropped from this instrument.
+                            }
+                            else
+                            {
+                                var metric = new Metric(instrument, temporality);
+                                this.metrics[index] = metric;
+                                this.metricStreamNames.Add(instrument.Name, true);
```

Comment:
What is the reason to make this a `Dictionary<string, bool>` rather than a `HashSet<string>`?

Reply:
There were (and still are) issues with HashSet performance in the .NET framework. So we just used Dictionary to be not affected by them.
Have to find the link to those discussions...